<a id="rna_hashing_top"></a>   

```{r}
orig_workspace <- ls() 

scrna_seq_sample_module_version <- "1.0.2" #20240227
source('/batchreporter/R/custom_functions.R')
```

The following metrics summarize the sequencing and alignment by Sample ID after hto-parsing and filtering for singlet cells.  


### Contents 

#### [Hash-Based Summary](#hto_scrna_summary)  
  - [Well Counts per HTO Barcode Plot](#well_hto_barcode)   
  - [Well Fraction per HTO Barcode Plot](#well_fraction_hto_barcode)  
  
#### [Count Stats per HTO Barcode](#count_stats_by_hto)  
  - [Read Counts per HTO Barcode Plot](#read_counts_hto)  
  - [UMI Counts per HTO Barcode Plot](#umi_counts_hto)  
  - [Gene Counts per HTO Barcode Plot](#gene_counts_hto)  
  
#### [Mitochondrial UMIs](#mito_umis_hto)  
  - [Fraction Mitochondrial Reads](#fraction_mito_reads_hto)  
  - [Fraction Mito UMIs by UMI Counts](#fraction_mito_reads_umi_hto_scatter)  
  - [Fraction Mito UMIs by Gene Counts](#fraction_mito_reads_genes_hto_scatter)  
  
#### [Data Quality UMAP](#rna_hash_data_quality_umap)  
#### [Biological Sex Verification Plots](#biological_sex_verification_plots)

<!-- create details section to collapse multiple no-output code blocks to save space -->
<details style="color: lightgray;">  
  <summary>Expand Code</summary> 

```{r include=FALSE}
stm("Starting scRNA Hashing Module (sample-level)")
```

```{css style_definitions, language="css"}
# Allow display of large images without shrinking to page width
<style>
  .superbigimage{
      overflow-x:scroll;
  }

  .superbigimage img{
     max-width: none;
  }

</style>
    
```

Check analysis dependencies  
```{r hto_rna_dependency_check, include=FALSE}
# check main rna input directory  
assertthat::assert_that(exists("in_rna"), 
                        msg = "Attempting to run rna quality module. Missing `in_rna` variable definition")

# check that we have h5 files provided. Can accept any type where 10x metadata has been injected
all_h5 <- list.files(path = in_rna, 
                               pattern = ".h5$", 
                               full.names = TRUE, recursive = TRUE)
assertthat::assert_that(length(all_h5) >0, 
                        msg = sprintf("Did not detect any .h5 files in input RNA directory %s", in_rna))

# check hash folder exists
assertthat::assert_that(dir.exists(in_hto), 
                        msg = "Attempting to run rna quality with hto module. Input `in_dir` does not contain 'hto' folder")

assertthat::assert_that(exists("in_key"),
                         msg = "Attempting to run quality with hto module. Missing `in_key` variable definition")
assertthat::assert_that(file.exists(in_key),
                         msg = sprintf("Attempting to rna quality with hto module. Input file `in_key` [%s] does not exist", in_key))

assertthat::assert_that(exists("in_config"),
                         msg = "Attempting to rna quality with hto module. Missing `in_config` variable definition")
assertthat::assert_that(file.exists(in_config),
                         msg = sprintf("Attempting to run rna quality with hto module. Input file `in_config` [%s] does not exist", in_config))

```


```{r hto_rna_read_config}
# Config
config_list<- load_config(in_config)

hash_key <- config_list$hash_key
sample_column_name <- config_list$sample_column_name

# Sample Key
df_key <- data.table::fread(in_key)

pools <- unique(df_key$PoolID)
n_pools <- length(pools)

wells <- unlist(strsplit(unique(df_key$WellID), split = ";"))
n_wells <- length(wells)

samples <- unique(df_key$SampleID)
n_samples <- length(samples)

```

Reading in metadata from h5 files  
```{r hto_rna_rna_metadata}
stm("Reading and merging all rna meta data")
rna_meta_list <- lapply(all_h5, H5weaver::read_h5_cell_meta)

# make sure column names are the same
col_list <- lapply(rna_meta_list, colnames)
all_cols_identical <- length(unique(col_list)) == 1
if(!all_cols_identical){
  all_columns <- unique(unlist(lapply(rna_meta_list, colnames)))
  common_columns <- Reduce(union, lapply(rna_meta_list, colnames))
  if(!all(all_columns %in% common_columns)){
    stm(sprintf("Warning: rna h5 files do not contain the same meta data columns. Keeping only the common columns. Removing columns %s.", 
        paste(setdiff(all_columns, common_columns), sep = ", ")))
  }
  rna_meta_list <- lapply(rna_meta_list, function(x){x[, common_columns]})
}

# merge metadata
rna_meta <- do.call(rbind, rna_meta_list)

# add metadata variables
rna_meta$fct_mito_umi <- rna_meta$n_mito_umi/rna_meta$n_umis
fct_mito_grp_cutoffs <- c(-Inf, 0.05, 0.10, 0.20, 0.30,Inf)
fct_mito_grp_labels <- c("0-0.05","0.05-0.10","0.10-0.20","0.20-0.30",">0.30")
rna_meta$fct_mito_group <- cut(rna_meta$fct_mito_umi, breaks =fct_mito_grp_cutoffs,
                           labels = fct_mito_grp_labels)
```

Read in Counts from H5 Files  
```{r hto_rna_rna_counts}
stm("Reading and merging all rna count matrices")

rna_count_list <- lapply(all_h5, H5weaver::read_h5_dgCMatrix, target = "matrix", 
                         feature_names = "id")
# make sure all matrices have same number of rows
if(!length(unique(sapply(rna_count_list, nrow)))==1){
  stop("RNA count matrixes have different numbers of rows")
} 

# make sure rows are in same order
row_order <- rownames(rna_count_list[[1]])
rna_count_list <- lapply(rna_count_list, function(x){x[row_order,]})

# make sure columns are in same order as metadata
order_check <- mapply(function(x, y){(all(x$barcodes==colnames(y)))}, rna_meta_list, rna_count_list)
if(!all(order_check)){
  # Reorder matrix columns to be consistent with metadata
  rna_count_list <- mapply( function(x, y){x[,y$barcodes]}, rna_count_list, rna_meta_list)
}

# merge
rna_counts <- do.call(cbind, rna_count_list)

featDF <- read_h5_feature_meta(all_h5[1])
```

Read in HTO files  
```{r hto_rna_hash_files}
stm("Reading in hto key for hashed RNA analysis")
# Read in all hto key to match hto names to barcodes
## hto_key <- system.file(file.path("reference",hash_key), package = "HTOparser")  # Parameterize this value
in_hto_key <- fread(hto_key, header = FALSE, col.names = c("hto_barcode","hto_name")) %>% 
  mutate(hto_order = as.numeric(gsub("HT","", hto_name))) %>% 
  mutate(hto_name = factor(hto_name, levels = hto_name[order(hto_order)])) %>% # use HT number value to reorder the HT levels
  select(-hto_order)

# Read in all hto metadata files, check expected numbers vs input well info
all_hto_meta <- list.files(path = in_hto, 
                           pattern = "hto_category_table.csv.gz$", 
                           full.names = TRUE, recursive = TRUE)
n_hto_meta <- length(all_hto_meta)
if(n_hto_meta == 0){
  stop(sprintf("No 'hto_category_table.csv.gz' files found in %s", in_hto))
} else if (n_hto_meta < n_wells){
  hto_meta_warn <- sprintf("Input number of 'hto_category_table.csv.gz' files (%s) is fewer than number of wells (%s) in sample key",
                           n_json, n_wells)
  warning(hto_meta_warn)
  hto_warning_list <- c(hto_warning_list, hto_meta_warn)
}
stm(paste0("IN HTO Metadata Files        :\n\t", paste(all_hto_meta, collapse = "\n\t")))
cat(paste0("IN HTO Metadata Files        :\n\t", paste(all_hto_meta, collapse = "\n\t"))) 

stm("Reading in hto category metadata for hashed RNA analysis")
hto_meta_list <- lapply(all_hto_meta, fread)
hto_meta_wells <- gsub("_.*","",basename(all_hto_meta))
hto_meta_list <- mapply(function(x,y){x$well_id <- y; x}, hto_meta_list, hto_meta_wells, SIMPLIFY = F)
hto_meta <- do.call(rbind, hto_meta_list)
rm("hto_meta_list")

```

Add in HTO Metadata  
```{r hto_rna_hto_meta_format}
stm("Formatting in hto category metadata for hashed RNA analysis")
hto_meta[in_hto_key, on = 'hto_barcode', hto_name := i.hto_name]  # merge in the hto names
hto_meta[ , pool_id := gsub("C\\dW\\d","", well_id)]
hto_meta[ , sample_hto:= sprintf("%s\n%s", hto_name, get(sample_column_name))]
hto_meta[ , sample_hto_pool:= sprintf("%s\n%s%s", hto_name, get(sample_column_name), pool_id)]
hto_meta[ , hto_order:=  as.numeric(gsub("HT","", hto_name))]
hto_meta[ , sample_hto_pool:=  factor(sample_hto_pool, levels = unique(sample_hto_pool[order(pool_id, hto_order)]))]
hto_meta[ , hto_order:= NULL]
hto_meta[ , hto_category:= factor(hto_category, levels = c("no_hash", "singlet", "doublet", "multiplet"))]

# hto_meta_singlet <- subset(hto_meta, hto_category=="singlet")

```

```{r hto_rna_merge}
stm("Merging hto category metadata with RNA metadata")

rna_hto_meta <- rna_meta %>%
    left_join(hto_meta, by = c("original_barcodes"="cell_barcode", "well_id" = "well_id","pool_id"="pool_id"))

hto_meta_singlet <- subset(rna_hto_meta, hto_category=="singlet")
hto_meta_singlet <- as.data.table(hto_meta_singlet)
hto_meta_singlet_list <- split(hto_meta_singlet, by = sample_column_name)
  
```

Sample singlet cells from each well and sample to generate umaps  
```{r hto_rna_sampled_cells}
n_cells_sample <- 2000

stm(sprintf("Sampling %s cells per sample and well (or all cells if fewer)", n_cells_sample))

set.seed(3)
sample_index_list <- lapply(hto_meta_singlet_list, function(x){
  sample_size = min(n_cells_sample, nrow(x))
  sort(sample(1:nrow(x), size = sample_size, replace = FALSE))
})

n_files <- length(sample_index_list)
rna_meta_list_sampled <- lapply(1:n_files, function(x){
 hto_meta_singlet_list[[x]][sample_index_list[[x]],]
})  

rna_count_list_sampled <- lapply(1:n_files, function(x){
 rna_counts[, rna_meta_list_sampled[[x]]$barcodes]
})

rna_meta_sampled <- do.call(rbind, rna_meta_list_sampled)
rna_meta_sampled$fct_mito_umi <- rna_meta_sampled$n_mito_umi/rna_meta_sampled$n_umis
rna_meta_sampled$fct_mito_group <- cut(rna_meta_sampled$fct_mito_umi, breaks =fct_mito_grp_cutoffs,
                           labels = fct_mito_grp_labels)
rna_counts_sampled <- do.call(cbind, rna_count_list_sampled)

rm(list=c("rna_count_list", "rna_count_list_sampled","rna_meta_list"))

```

Logic Check sample sheet value formats against data value formats
```{r hto_rna_check_ss_format}
all_pools_same <- identical(sort(pools), sort(unique(hto_meta$pool_id)))
if(!all_pools_same){
  intersect_pool <- intersect(pools, unique(hto_meta$pool_id))
  if(length(intersect_pool) == 0){
    stop(sprintf("Pool ID values in sample sheet [%s] do not match pool id values in data [%s].",
                 paste(pools, collapse = ";"), 
                 paste(sort(unique(hto_meta$pool_id)),collapse = ";")))
  }
  diff_sheet_pool <- setdiff(pools, unique(hto_meta$pool_id))
  if(length(diff_sheet_pool) >1){
    stm(sprintf("Warning: The following Pool values in sample sheet [%s] are not found in the hash tag data", 
        paste(diff_sheet_pool, collapse = ";")))
  }
  diff_data_pool <- setdiff( unique(hto_meta$pool_id),pools)
    if(length(diff_data_pool) >1){
      stm(sprintf("Warning: The following Pool values in the hash tag data [%s] are not found in the sample sheet.", 
        paste(diff_data_pool, collapse = ";")))
  }
}

stm(sprintf("Sampling %s cells per sample and well (or all cells if fewer)", n_cells_sample))

set.seed(3)
sample_index_list <- lapply(hto_meta_singlet_list, function(x){
  sample_size = min(n_cells_sample, nrow(x))
  sort(sample(1:nrow(x), size = sample_size, replace = FALSE))
})

n_files <- length(sample_index_list)
rna_meta_list_sampled <- lapply(1:n_files, function(x){
 hto_meta_singlet_list[[x]][sample_index_list[[x]],]
})  

rna_count_list_sampled <- lapply(1:n_files, function(x){
 rna_counts[, rna_meta_list_sampled[[x]]$barcodes]
})

rna_meta_sampled <- do.call(rbind, rna_meta_list_sampled)
rna_meta_sampled$fct_mito_umi <- rna_meta_sampled$n_mito_umi/rna_meta_sampled$n_umis
rna_meta_sampled$fct_mito_group <- cut(rna_meta_sampled$fct_mito_umi, breaks =fct_mito_grp_cutoffs,
                           labels = fct_mito_grp_labels)
rna_counts_sampled <- do.call(cbind, rna_count_list_sampled)

rm(list=c("rna_count_list", "rna_count_list_sampled","rna_meta_list"))

```

</details>  
  
<a id="hto_scrna_summary"></a>  

### Hash-Based Summary  

```{r hto_rna_hash_singlet_summary}
stm("Output pool based summary table")
hto_singlet_summary <- hto_meta_singlet[ ,.(n_singlet_cells = .N,
                             median_reads = median(n_reads),
                             median_umis = median(n_umis),
                             median_genes = median(n_genes)), 
                             by = setNames(list(pool_id, hto_name, hto_barcode, get(sample_column_name)), c('pool_id', 'hto_name','hto_barcode', sample_column_name))]
                          # by = .(pool_id, hto_name, hto_barcode, get(sample_column_name))]
setorder(hto_singlet_summary, pool_id, hto_name)
setcolorder(hto_singlet_summary, c("pool_id",sample_column_name, "hto_name", "hto_barcode", "n_singlet_cells", 
                            "median_reads", "median_umis", "median_genes"))

qc_table(hto_singlet_summary)

```

[Return to Contents](#rna_hashing_top)  

<a id="well_hto_barcode"></a> 

#### Well Count per HTO Barcode Plot  

```{r scrna_Count_By_HTO_Barcode_Plot, class.output="superbigimage", results="asis"}
stm("Plotting well count per hto")

plot_list <- list()
for (i in seq_along(pools)){
  plot_list[[i]] <- qc_aligned_barplot_facet(meta = hto_meta_singlet[pool_id == pools[i],],
                   category_x = "sample_hto_pool",
                   category_y = "well_id",
                   category_name = "Well ID",
                   name_x = "HTO/Sample",
                   colorset_y = "varibow",
                   name_y = "Number of Cells",
                   facet_formula = formula("~pool_id"), nrow = 1, scales ="free_x", drop = TRUE) 
}

g <- cowplot::plot_grid(plotlist = plot_list, ncol = 2)
temp_figwidth <- max(0.6*n_wells + 1*n_pools + 3, 8)
temp_figheight <- ceiling(n_pools/2)*4 + 0.4
make_subchunk(g, subchunk_name = "scrna_sample_hto_count_subchunk", 
              chunk_opt_list = list(fig.height = temp_figheight, fig.width = temp_figwidth, warning = FALSE, class.output = "superbigimage"), 
              quiet_knit = TRUE)

```

[Return to Contents](#rna_hashing_top)  
  
<a id="well_fraction_hto_barcode"></a> 

#### Well Fraction per HTO Barcode Plot  
```{r scrna_sampleFraction_By_HTO_Barcode_Plot, class.output="superbigimage", results="asis"}
stm("Plotting well fraction per hto")

g <- qc_stacked_barplot_facet(meta = hto_meta_singlet,
                   category_x = "sample_hto_pool",
                   category_y = "well_id",
                   category_name = "Well ID",
                   name_x = "HTO/Sample",
                   colorset_y = "varibow",
                   name_y = "Fraction of Cells",
                   as_fraction = TRUE , 
                   facet_formula = formula("~pool_id"), nrow = 1, scales ="free_x", drop = TRUE)

temp_figwidth <- max(0.3*n_wells + 0.5*n_pools + 1.5, 4)
temp_figheight <- ceiling(n_pools/2)*4 + 0.4
make_subchunk(g, subchunk_name = "scrna_sample_fraction_subchunk", 
              chunk_opt_list = list(fig.height = temp_figheight, fig.width = temp_figwidth, warning = FALSE), 
              quiet_knit = TRUE)

```

[Return to Contents](#rna_hashing_top)  

<a id="count_stats_by_hto"></a> 

### Count Stats per HTO Barcode  

<a id="read_counts_hto"></a> 
```{r}
dummy_plot <- function(msg, plottitle="", size = 8){
  ggplot(data.frame(1:2,1:2)) + 
        geom_text(x= 0.5,y =0.5,label=msg) +
        ggtitle(plottitle) +
       theme(text = element_text(size = size))

}
```
#### Read Counts by HTO Category and Barcode  
```{r hto_rna_read_counts, results="asis"}
# Reads per hto cat
stm("Generating sample read count violin plots")
category_reads_violins <- qc_violin_plot(rna_hto_meta,
                                         category_x = "hto_category",
                                         name_x = "HTO Category",
                                         column_y = "n_reads",
                                         name_y = "N Reads per Cell",
                                         fill = "dodgerblue")

# Reads per hto plot
g_read <- qc_violin_plot(hto_meta_singlet,
                        category_x = "sample_hto_pool",
                        name_x = "HTO/Sample (singlets)",
                        column_y = "n_reads",
                        name_y = "N Reads per Cell, Singlets",
                        log_y = TRUE,
                        fill = "dodgerblue") +
  ggtitle("Reads per Cell")
# g_read

reads_violin_list <- list(category_reads_violins, 
                          g_read)
g_grid <- plot_grid(plotlist = reads_violin_list,
          ncol = 2, rel_widths = c(1, 1+n_samples/4),
          nrow = 1, align = "h")
          
                    
temp_figwidth = 3 + n_samples*0.4
temp_figheight = 4

make_subchunk(g_grid, subchunk_name = "scrna_sample_counts_violin_subchunk", 
              chunk_opt_list = list(fig.height = temp_figheight, fig.width = temp_figwidth, 
                                    warning = FALSE), 
              quiet_knit = TRUE)
```
<details>  
  <summary>Expand plots of reads/cell distributions by hto and well </summary>  

```{r Read Count By HTO and Well Violin Plot, class.output="superbigimage", results="asis"}
stm("Plotting read count violin plots per hto and well")  

all_sample_hto_pool <- unique(sprintf("%s\n%s%s",df_key$HashTag,df_key$SampleID,  df_key$PoolID))

plot_list <- list()
for(i in seq_along(all_sample_hto_pool)){
    temp_meta <- filter(hto_meta_singlet, sample_hto_pool == all_sample_hto_pool[i])
    if(nrow(temp_meta) > 0){
          g <- qc_violin_plot(temp_meta,
                        category_x = "well_id",
                        name_x = "Well",
                        column_y = "n_reads",
                        name_y = "N Reads per Cell",
                        fill = "dodgerblue") +
      theme(text = element_text(size =8)) +
      ggtitle(all_sample_hto_pool[i])
    } else {
      g <- dummy_plot(msg = "No singlet cells", 
                      plottitle = all_sample_hto_pool[i], 
                      size =8 )
    }

    plot_list[[i]] <- g
}

g_grid <- cowplot::plot_grid(plotlist = plot_list, align = "tblr", ncol = 6)

temp_figwidth <- 15
temp_figheight <- ceiling(n_samples/6)*4 + 0.4
make_subchunk(g_grid, subchunk_name = "reads_well_hto_subchunk", 
              chunk_opt_list = list(fig.height = temp_figheight, fig.width = temp_figwidth, warning = FALSE), 
              quiet_knit = TRUE)

```
</details>  
  
[Return to Contents](#rna_hashing_top)  

<a id="umi_counts_hto"></a> 

#### UMI Counts by HTO Category and Barcode  

```{r hto_rna_umi_counts, results="asis"}
# UMI per category plot
stm("Generating sample umi count violin plots")
category_umis_violins <- qc_violin_plot(rna_hto_meta,
                                         category_x = "hto_category",
                                         name_x = "HTO Category",
                                         column_y = "n_umis",
                                         name_y = "N UMIs per Cell",
                                         fill = "purple")


# UMI per cell plot
g_umi <- qc_violin_plot(hto_meta_singlet,
                        category_x = "sample_hto_pool",
                        name_x = "HTO/Sample (singlets)",
                        column_y = "n_umis",
                        name_y = "N UMIs per Cell, Singlets",
                        fill = "purple") +
  ggtitle("UMIs per Cell")
# g_umi


umis_violin_list <- list(category_umis_violins, 
                          g_umi)
g_grid <- plot_grid(plotlist = umis_violin_list,
          ncol = 2, 
          # rel_widths = c(1, 4),
          rel_widths = c(1, 1+n_samples/4),
          nrow = 1, align = "h")
          
temp_figwidth = 3 + n_samples*0.4
temp_figheight = 4

make_subchunk(g_grid, subchunk_name = "scrna_sample_umi_violin_subchunk", 
              chunk_opt_list = list(fig.height = temp_figheight, fig.width = temp_figwidth, 
                                    warning = FALSE), 
              quiet_knit = TRUE)

```
<details>  
  <summary>Expand plots of UMIs/cell distributions per per well by hto </summary>  

```{r UMI Count By HTO and Well Violin Plot, class.output="superbigimage", results="asis"}
stm("Plotting umi count violin plots per hto and well")  

all_sample_hto_pool <- unique(sprintf("%s\n%s%s",df_key$HashTag,df_key$SampleID,  df_key$PoolID))

plot_list <- list()
for(i in seq_along(all_sample_hto_pool)){
    
    temp_meta <- filter(hto_meta_singlet, sample_hto_pool == all_sample_hto_pool[i])
    if(nrow(temp_meta) > 0) {
    g <- qc_violin_plot(temp_meta,
                        category_x = "well_id",
                        name_x = "Well",
                        column_y = "n_umis",
                        name_y = "N UMIs per Cell",
                        fill = "purple") +
      theme(text = element_text(size =8)) +
      ggtitle(all_sample_hto_pool[i])
    } else {
      g <- dummy_plot(msg = "No singlet cells", 
                      plottitle = all_sample_hto_pool[i], 
                      size = 8)
    }
    plot_list[[i]] <- g
}

g_grid <- cowplot::plot_grid(plotlist = plot_list, align = "tblr", ncol = 6)

temp_figwidth <- 15
temp_figheight <- ceiling(n_samples/6)*4 + 0.4
make_subchunk(g_grid, subchunk_name = "umis_well_hto_subchunk", 
              chunk_opt_list = list(fig.height = temp_figheight, fig.width = temp_figwidth, warning = FALSE), 
              quiet_knit = TRUE)
```
</details>    
  
[Return to Contents](#rna_hashing_top)  
  
<a id="gene_counts_hto"></a> 

#### Gene Counts by HTO Category and Barcode  

```{r hto_rna_gene_counts, results="asis"}
# Genes per category plot
category_genes_violins <- qc_violin_plot(rna_hto_meta,
                                         category_x = "hto_category",
                                         name_x = "HTO Category",
                                         column_y = "n_genes",
                                         name_y = "N Genes per Cell",
                                         fill = "orangered")  

# Genes per cell plot
g_genes <- qc_violin_plot(hto_meta_singlet,
                        category_x = "sample_hto_pool",
                        name_x = "HTO/Sample (singlets)",
                        column_y = "n_genes",
                        name_y = "N Genes per Cell, Singlets",
                        fill = "orangered") +
  ggtitle("Genes per Cell")


genes_violin_list <- list(category_genes_violins, 
                          g_genes)
g_grid <- plot_grid(plotlist = genes_violin_list,
          ncol = 2, 
          rel_widths = c(1, 1+n_samples/4),
          # rel_widths = c(1, 4),
          nrow = 1, align = "h")
          
temp_figwidth = 3 + n_samples*0.4
temp_figheight = 4

make_subchunk(g_grid, subchunk_name = "scrna_sample_genes_violin_subchunk", 
              chunk_opt_list = list(fig.height = temp_figheight, fig.width = temp_figwidth, 
                                    warning = FALSE), 
              quiet_knit = TRUE)

```

<details>  
  <summary>Expand plots of genes/cell distributions per per well by hto </summary>  

```{r Gene Count By HTO and Well Violin Plot, class.output="superbigimage", results="asis"}
stm("Plotting gene count violin plots per hto and well")  

all_sample_hto_pool <- unique(sprintf("%s\n%s%s",df_key$HashTag,df_key$SampleID,  df_key$PoolID))

plot_list <- list()
for(i in seq_along(all_sample_hto_pool)){
    
    temp_meta <- filter(hto_meta_singlet, sample_hto_pool == all_sample_hto_pool[i])
    if(nrow(temp_meta) > 0) {
    g <- qc_violin_plot(temp_meta,
                        category_x = "well_id",
                        name_x = "Well",
                        column_y = "n_genes",
                        name_y = "N Genes per Cell",
                        fill = "orangered") +
      theme(text = element_text(size = 8)) +
      ggtitle(all_sample_hto_pool[i])
    } else {
      g <- dummy_plot(msg = "No singlet cells", 
                plottitle = all_sample_hto_pool[i], 
                size = 8)
    }
    plot_list[[i]] <- g
}

g_grid <- cowplot::plot_grid(plotlist = plot_list, align = "tblr", ncol = 6)

temp_figwidth <- 15
temp_figheight <- ceiling(n_samples/6)*4 + 0.4
make_subchunk(g_grid, subchunk_name = "genes_well_hto_subchunk", 
              chunk_opt_list = list(fig.height = temp_figheight, fig.width = temp_figwidth, warning = FALSE), 
              quiet_knit = TRUE)
```
</details>  
  
<a id="mito_umis_hto" ></a>  

### Mitochondrial UMIs

<a id="fraction_mito_reads_hto" ></a>  

#### Fraction Mitochondrial UMIs  

```{r hto_rna_fct_mito_plot_nohash, results="asis"}
g_bar <- ggplot(hto_meta_singlet, aes(!!as.name(sample_column_name),  
                                      fill = factor(fct_mito_group, 
                                                    levels = rev(fct_mito_grp_labels)))) +
  geom_bar() +
  labs(fill = "Fraction Mitochondrial UMIs") +
  ylab("N Cells") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5))

g_pl <- plotly::ggplotly(g_bar) %>%
  layout(hovermode = 'compare')

tempwidth <- n_samples*0.4 + 2

make_subchunk(g_pl, subchunk_name = "scrna_sample_fraction_mito_subchunk", quiet_knit = T,
              chunk_opt_list = list(fig.width = tempwidth, fig.height = 6, echo = FALSE))

```

<a id="fraction_mito_reads_umi_hto_scatter" ></a>  

#### Fraction Mito UMIs by UMI Counts  

```{r hto_rna_mito_umis_vs_total, results="asis"}
# Reads per hto cat
stm("Generating mt umi vs umi count scatter plots by well")

# Reads per hto plot
g_mito_umi <- ggplot(hto_meta_singlet, aes(n_umis, fct_mito_umi)) +
    geom_point(color = "purple", alpha = 0.2) +
    xlab("UMI per Cell (log10 scale)") +
    ylab("Fraction Mitochondrial UMI") +
    scale_x_log10() +
    ggtitle("Fraction Mitochondrial UMI vs UMI per Cell") 

n_samples_plot <- length(unique(hto_meta_singlet[[sample_column_name]]))
ncols_plot <- 6
nrows_plot <- ceiling(n_samples_plot/ncols_plot)

# Reads per hto plot
g_mito_umi_sample <- g_mito_umi +
    facet_wrap(as.formula(paste0("~", sample_column_name)), 
               ncol = ncols_plot)

temp_figwidth <- 2*min(n_samples_plot, ncols_plot)+0.3
temp_figheight <- 2*nrows_plot + 0.3

make_subchunk(g_mito_umi_sample, subchunk_name = "scrna_sample_fraction_mito_umis_sample_subchunk", quiet_knit = T,
              chunk_opt_list = list(fig.width = temp_figwidth, fig.height = temp_figheight, echo = FALSE))

```

<details>  
  <summary> Expand Per-Well Plot </summary>  

```{r hto_rna_mito_umis_vs_umi_well, results="asis"}
# Reads per hto cat
stm("Generating mt umi vs umi count scatter plots per hto and well")

all_sample_hto_pool <- unique(sprintf("%s\n%s%s",df_key$HashTag,df_key$SampleID,  df_key$PoolID))

plot_list <- list()
for(i in seq_along(all_sample_hto_pool)){
    temp_meta <- hto_meta_singlet %>%
        dplyr::filter(sample_hto_pool == all_sample_hto_pool[i])
    if(nrow(temp_meta) > 0){
        g <- ggplot(temp_meta, aes(n_umis, fct_mito_umi)) +
          geom_point(color = "purple", alpha = 0.2) +
          xlab("UMI per Cell (log10 scale)") +
          ylab("Fraction Mitochondrial UMI") +
          scale_x_log10() +
          facet_grid( ~ well_id ) +
          ggtitle(all_sample_hto_pool[i])
    } else {
      g <- dummy_plot(msg = "No singlet cells", 
                      plottitle = all_sample_hto_pool[i])
    }

    plot_list[[i]] <- g
}

g_grid <- cowplot::plot_grid(plotlist = plot_list, align = "tblr", ncol = 1)

temp_figwidth <- 15
temp_figheight <- n_samples*4 + 0.4
make_subchunk(g_grid, subchunk_name = "fctmito_umiscatter_well_hto_subchunk", 
              chunk_opt_list = list(fig.height = temp_figheight, fig.width = temp_figwidth, warning = FALSE), 
              quiet_knit = TRUE)

```
</details>   
  
[Return to Contents](#rna_hashing_top)

<a id="fraction_mito_reads_genes_hto_scatter" ></a>  

#### Fraction Mito UMIs by Gene Counts  
```{r hto_rna_mito_umis_vs_genes, results="asis"}
# Genes per hto cat
stm("Generating mt umi vs gene count scatter plots by well")

# Genes per hto plot
g_mito_gene <- ggplot(hto_meta_singlet, aes(n_genes, fct_mito_umi)) +
    geom_point(color = "orangered", alpha = 0.2) +
    xlab("Genes per Cell (log10 scale)") +
    ylab("Fraction Mitochondrial UMI") +
    scale_x_log10() +
    ggtitle("Fraction Mitochondrial UMI vs Genes per Cell") 

n_samples_plot <- length(unique(hto_meta_singlet[[sample_column_name]]))
ncols_plot <- 6
nrows_plot <- ceiling(n_samples_plot/ncols_plot)

# Reads per hto plot
g_mito_genes_sample <- g_mito_gene +
    facet_wrap(as.formula(paste0("~", sample_column_name)), 
               ncol = ncols_plot)

temp_figwidth <- 2*min(n_samples_plot, ncols_plot)+0.3
temp_figheight <- 2*nrows_plot + 0.3

make_subchunk(g_mito_genes_sample, subchunk_name = "scrna_sample_fraction_mito_genes_sample_subchunk", quiet_knit = T,
              chunk_opt_list = list(fig.width = temp_figwidth, fig.height = temp_figheight, echo = FALSE))

```

<details>  
  <summary> Expand Per-Well Plot </summary>  
```{r hto_rna_mito_umis_vs_genes_well, results="asis"}
# Reads per hto cat
stm("Generating mt umi vs gene count scatter plots per hto and well")

all_sample_hto_pool <- unique(sprintf("%s\n%s%s",df_key$HashTag,df_key$SampleID,  df_key$PoolID))

plot_list <- list()
for(i in seq_along(all_sample_hto_pool)){
    temp_meta <- hto_meta_singlet %>%
        dplyr::filter(sample_hto_pool == all_sample_hto_pool[i])
    if(nrow(temp_meta) > 0){
      g <- ggplot(temp_meta, aes(n_genes, fct_mito_umi)) +
        geom_point(color = "orangered", alpha = 0.2) +
        xlab("UMI per Cell (log10 scale)") +
        ylab("Fraction Mitochondrial UMI") +
        scale_x_log10() +
        facet_grid( ~ well_id ) +
      ggtitle(all_sample_hto_pool[i])
    } else {
      g <- dummy_plot("No singlet cells", plottitle = all_sample_hto_pool[i])
    }
    plot_list[[i]] <- g
}

g_grid <- cowplot::plot_grid(plotlist = plot_list, align = "tblr", ncol = 1)

temp_figwidth <- 15
temp_figheight <- n_samples*4 + 0.4
make_subchunk(g_grid, subchunk_name = "fctmito_genescatter_well_hto_subchunk", 
              chunk_opt_list = list(fig.height = temp_figheight, fig.width = temp_figwidth, warning = FALSE), 
              quiet_knit = TRUE)

```

</details>   

[Return to Contents](#rna_hashing_top)
  
<a id="rna_hash_data_quality_umap"></a>  

### Data Quality UMAP  
<details style="color: lightgray;">  
  <summary> Expand Code </summary>  

```{r hto_rna_seurat_normalize}
# Create Seurat object
stm("Creating Seurat object from merged data for hashed RNA umap")
merged_so <- Seurat::CreateSeuratObject(counts = rna_counts_sampled)

# Normalize data
stm("Normalizing data for hashed RNA umap")
merged_so <- Seurat::NormalizeData(object = merged_so,
                                   normalization.method = "LogNormalize",
                                   scale.factor = 10000,
                                   margin = 1)

```

```{r hto_rna_PCA}
pc_dims <- min(ncol(merged_so) - 1, 50)
# suppressWarnings(future::plan("multiprocess", workers = avail_workers))

stm("Finding Variable Features for hashed RNA umap")
merged_so <- Seurat::FindVariableFeatures(object = merged_so)

stm("Scaling Data for hashed RNA umap")
merged_so <- Seurat::ScaleData(object = merged_so, verbose = FALSE)

stm("Running PCA for hashed RNA umap")
merged_so <- Seurat::RunPCA(object = merged_so, npcs = pc_dims, verbose = FALSE)

```

```{r hto_rna_jackstraw_pc_selection, message=FALSE}
# suppressWarnings(future::plan("multiprocess", workers = avail_workers))

stm("Determining dimensionality via jackstraw method for hashed RNA umap")

labels_order <- hto_meta_singlet[,get(sample_column_name)][match(colnames(merged_so),hto_meta_singlet$barcodes)]
names(labels_order) <- colnames(merged_so)

jackstraw_cells <- sample_cells(labels_order, 100, seed = 3030)

jackstraw_so <- merged_so[,jackstraw_cells]

jackstraw_so <- Seurat::JackStraw(object = jackstraw_so,
                         dims = pc_dims,
                         num.replicate = 50,  
                         verbose = FALSE)
jackstraw_so <- Seurat::ScoreJackStraw(object = jackstraw_so,
                              dims = 1:pc_dims)

pc_pvals <- jackstraw_so@reductions$pca@jackstraw$overall.p.values[,2]
good_pcs <- pc_pvals < 0.05

nPC <- sum(good_pcs)

pc_var <- Stdev(merged_so, reduction = "pca")^2
total_var <- merged_so@reductions$pca@misc$total.variance
var_selected_pc <- sum(pc_var[good_pcs])/total_var
cumvar_string <- sprintf(fmt = "%.1f", var_selected_pc*100)

stm(sprintf("Selected %s significant pcs via JackStraw, %s%% explained variation", nPC, cumvar_string))

rm(jackstraw_so)
```

```{r hto_rna_merged_umap}
# Run UMAP
# suppressWarnings(future::plan("multiprocess", workers = avail_workers))
stm(sprintf("Running UMAP on selected coordinates for hashed RNA umap"))

merged_so <- Seurat::RunUMAP(merged_so,
                                   dims = c(1:50)[good_pcs],
                                   umap.method = "uwot",
                                   seed.use = 3,
                                   verbose = FALSE)

```


```{r hto_rna_umap_cleanup}
# Get UMAP coordinates
umapDF <- merged_so[["umap"]]@cell.embeddings %>%
          as.data.frame() %>%
          dplyr::rename(UMAP_1_merged = UMAP_1, UMAP_2_merged = UMAP_2) %>%
          tibble::rownames_to_column(var = "barcodes")
umapDF <- merge(umapDF, hto_meta_singlet, by = "barcodes")
rownames(umapDF) <- umapDF$barcodes

umapDF <- umapDF[rownames(merged_so@"meta.data"), , drop = F]

```

</details>  
<!-- end of details section collapsing multiple no-output code blocks to save space -->  

Batch-level UMAP using `r {nPC}` principal components (`r {cumvar_string}`% explained variance) selected by jackstraw. 
Singlet cells only, `r {n_cells_sample}` randomly sampled cells per Sample ID.

```{r umap_quality, results="asis", fig.height=30, fig.width=20, class.output="superbigimage"}
stm("Plotting Batch UMAP by Data Quality Metrics for hashed RNA")

# Gene scaling
max_genes <- max(8000, hto_meta_singlet$n_genes)
max_umi <- max(60000, hto_meta_singlet$n_umis)
point_size <- 0.2


# Cell Types
g_base <- ggplot(umapDF, aes(UMAP_1_merged, UMAP_2_merged))

# fraction mitochondrial umi
stm("Plotting Fraction Mito UMAP")
g3 <- plot_umap_report(df = umapDF,
                 x_col="UMAP_1_merged",
                 x_lab = "UMAP 1",
                 y_col = "UMAP_2_merged",
                 y_lab = "UMAP 2",
                 title = "Fraction Mitochondrial UMIs",
                 point_size = point_size,
                 color_col = "fct_mito_umi",
                 scale_color_fun = scale_color_fct_mito)

# N Genes
stm("Plotting N Genes UMAP")
g4 <- plot_umap_report(df = umapDF,
                 x_col="UMAP_1_merged",
                 x_lab = "UMAP 1",
                 y_col = "UMAP_2_merged",
                 y_lab = "UMAP 2",
                 title = "N Genes",
                 point_size = point_size,
                 color_col = "n_genes",
                 scale_color_fun = scale_color_genes(max_genes)
)

# N UMIs
stm("Plotting N UMIs UMAP")
g5 <- plot_umap_report(df = umapDF,
                 x_col="UMAP_1_merged",
                 x_lab = "UMAP 1",
                 y_col = "UMAP_2_merged",
                 y_lab = "UMAP 2",
                 title = "N UMIs",
                 point_size = point_size,
                 color_col = "n_umis",
                 scale_color_fun = scale_color_umis(max_umi)
)


# Wells
stm("Plotting Well UMAP")
cols_well <- H5weaver::varibow(n_colors = length(unique(umapDF$well_id)))

scale_color_well <- function(...){
    scale_color_manual(values = cols_well, ...)
}
g6 <- plot_umap_report(df = umapDF,
                 x_col="UMAP_1_merged",
                 x_lab = "UMAP 1",
                 y_col = "UMAP_2_merged",
                 y_lab = "UMAP 2",
                 title = "Well ID",
                 point_size = point_size,
                 color_col = "well_id",
                 scale_color_fun = scale_color_well
) +
  guides(colour = guide_legend(override.aes = list(size = 3)))

# Sample
stm("Plotting Sample ID UMAP")
cols_sample <- H5weaver::varibow(n_colors = length(unique(umapDF[,sample_column_name])))
scale_color_sample <- function(...){
    scale_color_manual(values = cols_sample, ...)
}

g7 <- plot_umap_report(df = umapDF,
                 x_col="UMAP_1_merged",
                 x_lab = "UMAP 1",
                 y_col = "UMAP_2_merged",
                 y_lab = "UMAP 2",
                 title = "Sample ID",
                 point_size = point_size,
                 color_col = sample_column_name,
                 scale_color_fun = scale_color_sample
) +
  guides(colour = guide_legend(override.aes = list(size = 3)))

aligned_plots <- cowplot::align_plots(g3, g4, g5, g6, g7, align = "hv", axis = "tblr")  # uniform plot sizing

cowplot::plot_grid(aligned_plots[[1]],
                   aligned_plots[[2]],
                   aligned_plots[[3]],
                   aligned_plots[[4]],
                   aligned_plots[[5]],
                   ncol = 2)

```

[Return to Contents](#rna_hashing_top) 

<a id="biological_sex_verification_plots"></a>  

### Biological Sex Verification Plots
#### Gene selected to identify Female samples - XIST : ENSG00000229807
#### Gene selected to identify Male samples   - RPS4Y1 : ENSG00000129824

```{r sex verification plots function definitions}
make_dotplot <- function(summary_df, exp = 'avg.exp.scaled'){
    ggplot(summary_df, aes(id, features.plot, size = pct.exp, color = !!as.name(exp))) +
        geom_point() +
        scale_color_viridis_c(option = 'turbo') +
        theme_bw(base_size = 16) + 
        scale_y_discrete(position = "right") +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
        facet_grid(~subject.biologicalSex, scales = 'free_x', space = 'free') 
}

make_box <- function(summary_df, y_col = 'avg.exp', ...){
    ggplot(summary_df, aes(subject.biologicalSex, !!as.name(y_col), color = subject.biologicalSex, label = id)) +
        geom_point(position = position_jitter(width = 0.3, height = 0)) + 
        geom_boxplot(alpha = 0, color = 'black') +
        theme_bw(base_size = 16) +
        facet_grid(~features.plot, scales = 'free_x', space = 'free', ...) + 
        theme(axis.text.x = element_text(angle = 60), axis.title.x = element_text(margin = margin(t = 30), size = 8))
}

```


```{r sex verification plot formatting}
stm('Formating data for sex verification plots')

rna_meta_sampled <- as.data.frame(rna_meta_sampled)
rownames(rna_meta_sampled) <- rna_meta_sampled$barcodes
merged_so@meta.data <- rna_meta_sampled

if('subject.biologicalSex' %in% colnames(df_key)) {
    sample_sex = TRUE
    sex_meta = df_key[,c('SampleID','subject.biologicalSex')]
    sex_meta$subject.biologicalSex <- ifelse(! sex_meta$subject.biologicalSex %in% c('Male','Female'), NA, sex_meta$subject.biologicalSex)
    if(!all(sex_meta$subject.biologicalSex %in% c('Male','Female',NA))) {
           stm("subject.biologicalSex values expected to be Male, Female or NA. Please check values.")
           sample_sex = FALSE}
    } else 
    {
    sample_sex = FALSE
    }

Idents(merged_so) <- 'pbmc_sample_id'
samples <- levels(merged_so)
DefaultAssay(merged_so) <- 'RNA'

if(sample_sex) {

    assert_that('SampleID' %in% colnames(sex_meta))
    assert_that('subject.biologicalSex' %in% colnames(sex_meta)) 
    
    if(! all(samples %in% sex_meta$SampleID)) { 
        stm('Samples in batch do not match samples in sex metadata provided')
        stm('Ignoring sample sex metadata provided')
        sample_sex = FALSE
    }

}

gene_ids <- rownames(merged_so@assays$RNA@data)[1]

female_gene <- 'XIST'
male_gene <- 'RPS4Y1'

if(grepl('ENSG', gene_ids, fixed = TRUE)) { 
    female_gene <- 'ENSG00000129824'
    male_gene <- 'ENSG00000229807'
}

keep_genes <- c(male_gene, female_gene)

plot <- DotPlot(object = merged_so, features = keep_genes, group.by = 'pbmc_sample_id')
plot_data <- plot$data
rownames(plot_data) <- NULL

if(sample_sex) {
    plot_data_merged <- left_join(plot_data, sex_meta, by = c("id" = "SampleID")) 
} else {
        plot_data_merged <- plot_data
        plot_data_merged$subject.biologicalSex <- NA    
    }

ord <- filter(plot_data_merged, features.plot == female_gene)
order <- ord[order(ord$pct.exp, decreasing = FALSE), ]
order <- order$id

plot_data_merged <- plot_data_merged %>% mutate(id = factor(id, levels = order))

#plotdim(15,6)
p1 <- make_dotplot(plot_data_merged, exp = 'avg.exp.scaled') + 
ggtitle(label = paste0("Sex based gene expression in ", unique(merged_so$pool_id)))

if(sample_sex) {
    avg_exp_box <- make_box(plot_data_merged, y_col = 'avg.exp') +  
    ggtitle(label = paste0("Average exp of sex based markers in ", unique(merged_so$pool_id))) + 
    theme(plot.title = element_text(margin = margin(0,0,30,0))) 
    avg_plotly <- suppressWarnings(ggplotly(avg_exp_box, tooltip = 'id'))
    p2 <- avg_plotly
}

if(sample_sex) {
    pct_exp_box <- make_box(plot_data_merged, y_col = 'pct.exp')  +  
    ggtitle(label = paste0("Percent exp of sex based markers in ", unique(merged_so$pool_id))) + 
    theme(plot.title = element_text(margin = margin(0,0,30,0)))
    pct_plotly <- suppressWarnings(ggplotly(pct_exp_box, tooltip = 'id'))
    p3 <- pct_plotly
}

```

```{r sex_verification_dot_plot, results="asis", fig.height=8, fig.width=18, class.output="superbigimage"}

stm("Plotting Sex Verification Dot Plot")
print(p1)

```

```{r sex_verification_box_plots, results="asis", fig.width=18}

if(sample_sex) {
    boxplots = list(p2,p3)
    stm("Plotting Sex Verification Box Plots")
    htmltools::tagList(boxplots)
}

```

[Return to Contents](#rna_hashing_top)  

---  

scRNA seq report sample module v.`r {scrna_seq_sample_module_version}`, Lauren Okada  


```{r scrna_seq_sample_cleanup, include=FALSE}
module_vars <- setdiff(ls(), orig_workspace)
rm(list=module_vars)

gc()

```
